### PR TITLE
Add a scheme to find the place where an id was destroyed

### PR DESCRIPTION
**Pull request by oli-obk**
cc https://github.com/rust-lang/miri/issues/974

I'm not too happy with it, but since stacked borrows don't have access to the current call stack, I can't just report a warning as per https://github.com/rust-lang/miri/issues/797

We could add some global mutex that we can throw strings at and `step` will clear out that mutex and report warnings before moving the `statement_id` or the `block_id`, not sure how well that would work. For now I think this is sufficient

### CODE REVIEW

File: tests/compile-fail/stacked_borrows/track_id.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// compile-flags: -Zmiri-track-id=1372
+// do not run on anything but x86_64 linux, because minute changes can change the borrow stack ids
```
Comment (bjorn3) [inline tests/compile-fail/stacked_borrows/track_id.rs]: s/minute/minimal?

File: src/eval.rs [hunk 2]
```diff
@@ -27,6 +27,8 @@ pub struct MiriConfig {
     pub args: Vec<String>,
     /// The seed to use when non-determinism or randomness are required (e.g. ptr-to-int cast, `getrandom()`).
     pub seed: Option<u64>,
+    /// The stacked borrow id to report about
+    pub tracked_id: Option<PtrId>,
```
Comment (RalfJung) [inline src/eval.rs]: We never call these "borrow ID". The usual name is "Pointer tag". Yes, tags include all `PtrId` and `\bottom`, aka `Option<PtrId>`, but still -- the term "pointer ID" is entirely an implementation detail of Stacked Borrows and should not be used outside.

I propose `tracked_pointer_tag`.

File: src/bin/miri.rs [hunk 3]
```diff
@@ -172,6 +173,17 @@ fn main() {
                 arg if arg.starts_with("-Zmiri-env-exclude=") => {
                     excluded_env_vars.push(arg.trim_start_matches("-Zmiri-env-exclude=").to_owned());
                 },
+                arg if arg.starts_with("-Zmiri-track-id=") => {
```
Comment (RalfJung) [inline src/bin/miri.rs]: Please also add the new flag to the README.

File: src/stacked_borrows.rs [hunk 4]
```diff
@@ -270,6 +271,11 @@ impl<'tcx> Stack {
 
     /// Check if the given item is protected.
     fn check_protector(item: &Item, tag: Option<Tag>, global: &GlobalState) -> InterpResult<'tcx> {
+        if let Tag::Tagged(id) = item.tag {
+            if Some(id) == global.tracked_id {
+                throw_unsup!(Unsupported(format!("disabling item {:?} for tag {:?}", item, tag)));
```
Comment (RalfJung) [inline src/stacked_borrows.rs]: Uh, "unsup" really doesn't fit at all here...

File: tests/compile-fail/stacked_borrows/track_id.rs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+// compile-flags: -Zmiri-track-id=1372
```
Comment (RalfJung) [inline tests/compile-fail/stacked_borrows/track_id.rs]: This test will break any time *anything* in the libstd prelude changes. I don't think we should have to update Miri each time that happens.

I'd rather have no test than this test, TBH.

**Comment by RalfJung** [comment]
What a hack! But it's not like I have a better idea that's even remotely as simple, so, the basic approach seems fine to me.

File: src/stacked_borrows.rs (hunk 4 above)
Comment (oli-obk) [reply src/stacked_borrows.rs]: None of the others fit either. I chose one where I could specify a pure string without actually signalling any additional information by the name of the variant

File: tests/compile-fail/stacked_borrows/track_id.rs (hunk 5 above)
Comment (oli-obk) [reply tests/compile-fail/stacked_borrows/track_id.rs]: I can work on making miri run no-std tests again, then there'll be nothing changing the ids. But yea, for now I'll remove the test

File: src/stacked_borrows.rs (hunk 4 above)
Comment (RalfJung) [reply src/stacked_borrows.rs]: Please add a FIXME to use a `MachineError` variant (or so) then, once we added that.

Comment (RalfJung) [reply src/stacked_borrows.rs]: r=me with this comment and after squashing.

Comment (RalfJung) [reply src/stacked_borrows.rs]: With https://github.com/rust-lang/miri/pull/1092 we now have the `MachineStop` variant available, so could you use that instead?

Comment (RalfJung) [reply src/stacked_borrows.rs]: @oli-obk  This concern is still open.

Comment (oli-obk) [reply src/stacked_borrows.rs]: Yea I'm currently unable to build miri with the toolchain script. Getting some errors about libsyntax not found, will try again with nightly

Comment (RalfJung) [reply src/stacked_borrows.rs]: What's your version of `rustup-toolchain-install-master`? You might have to upgrade for it to properly support `rustc-dev`.

Comment (oli-obk) [reply src/stacked_borrows.rs]: That was it, thanks

Comment (oli-obk) [reply src/stacked_borrows.rs]: This gave me an idea wrt continuing execution, I hope you don't hate it too much :laughing: We may corrupt some state with it.

File: README.md [hunk 6]
```diff
@@ -171,6 +171,11 @@ Several `-Z` flags are relevant for Miri:
   sets this flag per default.
 * `-Zmir-emit-retag` controls whether `Retag` statements are emitted. Miri
   enables this per default because it is needed for validation.
+* `-Zmiri-track-pointer-tag` aborts interpretation with a backtrace when the
```
Comment (RalfJung) [inline README.md]:
```suggestion
* `-Zmiri-track-pointer-tag=<tag>` aborts interpretation with a backtrace when the
```

**Comment by RalfJung** [comment]
Uh... that's really nasty. And you're aborting in the middle of a Stacked Borrows operation. I don't think it is a good idea to go on executing -- as you said, that's basically a corrupt state, so any follow-up errors might be completely bogus.

I prefer your first scheme.

**Comment by RalfJung** [comment]
LGTM! r=me with the open suggestion applied.

**Pull request merged**